lib/auth/auth_with_roles.go

```diff
@@ -2039,21 +2067,40 @@ func (a *ServerWithRoles) UploadSessionRecording(r events.SessionRecording) erro
 }
 
 func (a *ServerWithRoles) GetSessionChunk(namespace string, sid session.ID, offsetBytes, maxBytes int) ([]byte, error) {
-	if err := a.sessionRecordingAction(namespace, sid, types.VerbRead); err != nil {
+	if err := a.actionForKindSession(namespace, types.VerbRead, sid); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	return a.alog.GetSessionChunk(namespace, sid, offsetBytes, maxBytes)
 }
 
 func (a *ServerWithRoles) GetSessionEvents(namespace string, sid session.ID, afterN int, includePrintEvents bool) ([]events.EventFields, error) {
-	if err := a.sessionRecordingAction(namespace, sid, types.VerbRead); err != nil {
+	if err := a.actionForKindSession(namespace, types.VerbRead, sid); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	return a.alog.GetSessionEvents(namespace, sid, afterN, includePrintEvents)
 }
 
+func (a *ServerWithRoles) findSessionEndEvent(namespace string, sid session.ID) (*apievents.SessionEnd, error) {
+	sessionEvents, err := a.alog.GetSessionEvents(namespace, sid, 0, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	for _, ef := range sessionEvents {
```

Comment:
I think here you can use `SearchEvents()` and pass in `session.end` string value in eventTypes ([]string) instead of iterating through all events. There should only be one `session.end` event per session.